File: MIDTERM-LAB-1/TOLENTINO_MIDTERM_LAB1.js
const fs = require('fs');
const readline = require('readline');
const path = require('path');

function splitCsvLine(line) {
  const result = [];
  let cur = '';
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (ch === '"') {
      if (inQuotes && line[i + 1] === '"') { cur += '"'; i++; } else { inQuotes = !inQuotes; }
      continue;
    }
    if (ch === ',' && !inQuotes) {
      result.push(cur);
      cur = '';
      continue;
    }
    cur += ch;
  }
  result.push(cur);
  return result;
}

function findColumn(headers, candidates) {
  for (const candidate of candidates) {
    for (let i = 0; i < headers.length; i++) {
      if (headers[i].trim().toLowerCase() === candidate.toLowerCase()) return i;
    }
  }
  return -1;
}

async function prompt(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(resolve => rl.question(question, ans => { rl.close(); resolve(ans); }));
}

async function getValidFilePath() {
  while (true) {
    const input = (await prompt('\nEnter dataset file path: ')).trim();
    let p = input;
    if (p.startsWith('"') && p.endsWith('"')) p = p.slice(1, -1);
    if (!fs.existsSync(p)) { console.log('[ERROR] File not found.'); continue; }
    try { fs.accessSync(p, fs.constants.R_OK); } catch (e) { console.log('[ERROR] File not readable.'); continue; }
    if (path.extname(p).toLowerCase() !== '.csv') { console.log('[ERROR] Not a .csv file.'); continue; }
    const firstLine = fs.readFileSync(p, { encoding: 'utf8' }).split(/\r?\n/)[0] || '';
    if (!firstLine.includes(',')) { console.log('[ERROR] File does not look like CSV.'); continue; }
    console.log('[OK] File validated successfully!');
    return p;
  }
}

function loadData(filePath) {
  const text = fs.readFileSync(filePath, { encoding: 'utf8' });
  const lines = text.split(/\r?\n/).filter(Boolean);
  if (lines.length === 0) throw new Error('CSV file is empty.');
  const headers = splitCsvLine(lines[0]);

  const titleIdx = findColumn(headers, ['title','game','name']);
  const salesIdx = findColumn(headers, ['total_sales','global_sales','sales']);
  if (titleIdx === -1) throw new Error('Could not find a title/game column in CSV.');
  if (salesIdx === -1) throw new Error('Could not find a sales column in CSV.');

  console.log('[INFO] Title column :', headers[titleIdx].trim());
  console.log('[INFO] Sales column :', headers[salesIdx].trim());

  const salesMap = new Map();
  let lineCount = 0, errorCount = 0;
  for (let i = 1; i < lines.length; i++) {
    lineCount++;
    try {
      const cols = splitCsvLine(lines[i]);
      if (cols.length <= Math.max(titleIdx, salesIdx)) continue;
      const title = cols[titleIdx].trim().replace(/"/g, '');
      const salesStr = cols[salesIdx].trim().replace(/"/g, '');
      if (!title || !salesStr) continue;
      const sales = parseFloat(salesStr);
      if (Number.isNaN(sales)) { errorCount++; continue; }
      salesMap.set(title, (salesMap.get(title) || 0) + sales);
    } catch (e) { errorCount++; }
  }

  console.log('[INFO] Rows processed :', lineCount);
  console.log('[INFO] Rows skipped   :', errorCount);

  const records = Array.from(salesMap.entries()).map(([title, totalSales]) => ({ title, totalSales }));
  records.sort((a,b) => b.totalSales - a.totalSales);
  return records;
}

async function main() {
  console.log('============================================================');
  console.log('   MIDTERM LAB 1 - Detect Low Performing Products');
  console.log('   Student: TOLENTINO | Prog2-9302');
  console.log('   Converted to Node.js');
  console.log('============================================================');

  const filePath = await getValidFilePath();
  try {
    console.log('\n[INFO] Loading dataset...');
    const records = loadData(filePath);
    console.log('[INFO] Unique products found:', records.length);
    if (records.length === 0) { console.log('[ERROR] No valid records found.'); return; }
    const totalSalesSum = records.reduce((s,r) => s + r.totalSales, 0);
    const average = totalSalesSum / records.length;

    console.log('\n============================================================');
    console.log(`  Dataset Average Total Sales: ${average.toFixed(2)} million units`);
    console.log('============================================================');

    const lowPerformers = records.filter(r => r.totalSales < average).sort((a,b) => a.totalSales - b.totalSales);

    console.log('\n============================================================');
    console.log('  FLAGGED LOW PERFORMING PRODUCTS');
    console.log(`  (Sales below average: ${average.toFixed(2)} M)`);
    console.log('============================================================');
    console.log(`${'Product Title'.padEnd(55)} ${'Sales (M)'.padStart(10)}   Status`);
    console.log('------------------------------------------------------------');
    for (const r of lowPerformers) {
      console.log(`${r.title.padEnd(55)} ${r.totalSales.toFixed(2).padStart(10)}   FLAGGED`);
    }

    console.log('\n============================================================');
    console.log('  SUMMARY');
    console.log('============================================================');
    console.log('  Total Products Analyzed  :', records.length);
    console.log('  Average Sales            :', average.toFixed(2), 'million units');
    console.log('  Low Performing (Flagged) :', lowPerformers.length);
    console.log('  Good Performing          :', records.length - lowPerformers.length);
    console.log('============================================================');
    console.log('\n[DONE] Inventory team can now review flagged products.\n');

  } catch (e) { console.log('[ERROR] Failed to process file:', e.message); }
}

if (require.main === module) main();

File: MIDTERM-LAB-1/TOLENTINO_MIDTERM_LAB1.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

public class TOLENTINO_MIDTERM_LAB1 {

    public static String getValidFilePath(Scanner scanner) {
        while (true) {
            System.out.print("\nEnter dataset file path: ");
            String path = scanner.nextLine().trim();

            if (path.startsWith("\"") && path.endsWith("\"")) {
                path = path.substring(1, path.length() - 1);
            }

            File file = new File(path);

            if (!file.exists()) {
                System.out.println("[ERROR] File not found. Please check the path and try again.");
                continue;
            }
            if (!file.canRead()) {
                System.out.println("[ERROR] File cannot be read. Check file permissions.");
                continue;
            }
            if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("[ERROR] File is not a CSV file. Please provide a .csv file.");
                continue;
            }

            try (BufferedReader br = new BufferedReader(new FileReader(file))) {
                String header = br.readLine();
                if (header == null || !header.contains(",")) {
                    System.out.println("[ERROR] File does not appear to be a valid CSV format.");
                    continue;
                }
                System.out.println("[OK] File validated successfully!");
                return path;
            } catch (IOException e) {
                System.out.println("[ERROR] Could not read file: " + e.getMessage());
            }
        }
    }

    public static int findColumn(String[] headers, String... candidates) {
        for (String candidate : candidates) {
            for (int i = 0; i < headers.length; i++) {
                if (headers[i].trim().toLowerCase().equals(candidate.toLowerCase())) {
                    return i;
                }
            }
        }
        return -1;
    }

    public static List<DataRecord> loadData(String filePath) throws IOException {
        Map<String, Double> salesMap = new HashMap<>();
        int titleIdx = -1;
        int salesIdx = -1;
        int lineCount = 0;
        int errorCount = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String headerLine = br.readLine();
            if (headerLine == null) throw new IOException("CSV file is empty.");
            String[] headers = headerLine.split(",", -1);

            titleIdx = findColumn(headers, "title", "Title", "game", "Game", "name", "Name");
            salesIdx = findColumn(headers, "total_sales", "Total_Sales", "global_sales",
                                  "Global_Sales", "sales", "Sales");

            if (titleIdx == -1) throw new IOException("Could not find a title/game column in CSV.");
            if (salesIdx == -1) throw new IOException("Could not find a sales column in CSV.");

            System.out.println("[INFO] Title column : " + headers[titleIdx].trim());
            System.out.println("[INFO] Sales column : " + headers[salesIdx].trim());

            String line;
            while ((line = br.readLine()) != null) {
                lineCount++;
                try {
                    String[] cols = line.split(",", -1);
                    if (cols.length <= Math.max(titleIdx, salesIdx)) continue;

                    String title = cols[titleIdx].trim().replace("\"", "");
                    String salesStr = cols[salesIdx].trim().replace("\"", "");

                    if (title.isEmpty() || salesStr.isEmpty()) continue;

                    double sales = Double.parseDouble(salesStr);
                    salesMap.put(title, salesMap.getOrDefault(title, 0.0) + sales);

                } catch (NumberFormatException e) {
                    errorCount++;
                }
            }
        }

        System.out.println("[INFO] Rows processed : " + lineCount);
        System.out.println("[INFO] Rows skipped   : " + errorCount);

        List<DataRecord> records = new ArrayList<>();
        for (Map.Entry<String, Double> entry : salesMap.entrySet()) {
            records.add(new DataRecord(entry.getKey(), entry.getValue()));
        }
        records.sort((a, b) -> Double.compare(b.getTotalSales(), a.getTotalSales()));
        return records;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("============================================================");
        System.out.println("   MIDTERM LAB 1 - Detect Low Performing Products");
        System.out.println("   Student: TOLENTINO | Prog2-9302");
        System.out.println("   University of Perpetual Help - Molino Campus");
        System.out.println("============================================================");

        String filePath = getValidFilePath(scanner);

        try {
            System.out.println("\n[INFO] Loading dataset...");
            List<DataRecord> records = loadData(filePath);
            System.out.println("[INFO] Unique products found: " + records.size());

            if (records.isEmpty()) {
                System.out.println("[ERROR] No valid records found in the dataset.");
                return;
            }

            double totalSalesSum = 0;
            for (DataRecord r : records) {
                totalSalesSum += r.getTotalSales();
            }
            double average = totalSalesSum / records.size();

            System.out.println("\n============================================================");
            System.out.printf("  Dataset Average Total Sales: %.2f million units%n", average);
            System.out.println("============================================================");

            List<DataRecord> lowPerformers = new ArrayList<>();
            for (DataRecord r : records) {
                if (r.getTotalSales() < average) {
                    lowPerformers.add(r);
                }
            }
            lowPerformers.sort((a, b) -> Double.compare(a.getTotalSales(), b.getTotalSales()));

            System.out.println("\n============================================================");
            System.out.println("  FLAGGED LOW PERFORMING PRODUCTS");
            System.out.printf("  (Sales below average: %.2f M)%n", average);
            System.out.println("============================================================");
            System.out.printf("%-55s %10s   %s%n", "Product Title", "Sales (M)", "Status");
            System.out.println("------------------------------------------------------------");

            for (DataRecord r : lowPerformers) {
                System.out.printf("%s   %s%n", r.toString(), "FLAGGED");
            }

            System.out.println("\n============================================================");
            System.out.println("  SUMMARY");
            System.out.println("============================================================");
            System.out.println("  Total Products Analyzed  : " + records.size());
            System.out.printf("  Average Sales            : %.2f million units%n", average);
            System.out.println("  Low Performing (Flagged) : " + lowPerformers.size());
            System.out.println("  Good Performing          : " + (records.size() - lowPerformers.size()));
            System.out.println("============================================================");
            System.out.println("\n[DONE] Inventory team can now review flagged products.\n");

        } catch (IOException e) {
            System.out.println("[ERROR] Failed to process file: " + e.getMessage());
        }

        scanner.close();
    }
}
